{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-10-25T04:35:04.263822Z",
     "iopub.status.busy": "2020-10-25T04:35:04.263587Z",
     "iopub.status.idle": "2020-10-25T04:35:04.279404Z",
     "shell.execute_reply": "2020-10-25T04:35:04.278886Z",
     "shell.execute_reply.started": "2020-10-25T04:35:04.263800Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.Alphabet import IUPAC\n",
    "from pathlib import Path\n",
    "from typing import Union\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-10-25T04:49:48.933964Z",
     "iopub.status.busy": "2020-10-25T04:49:48.933733Z",
     "iopub.status.idle": "2020-10-25T04:49:48.942548Z",
     "shell.execute_reply": "2020-10-25T04:49:48.941179Z",
     "shell.execute_reply.started": "2020-10-25T04:49:48.933941Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_primers(dna, start, end, output_file: Union[str, Path],\n",
    "                     codon='NNS', length_primer=15, tm=None, return_df=False):\n",
    "    '''\n",
    "    Generate primers for saturation mutagenesis.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dna : string\n",
    "        DNA sequence containing the protein of study. \n",
    "        The DNA sequence should also contain at least 15 base pairs before the \n",
    "        starting ATG and 15 base pairs after the stop codon.\n",
    "\n",
    "    start : string\n",
    "        Beginning of the DNA sequence that will be mutageneized.\n",
    "        For example, if you will start mutating the first methionine, copy a few more dna bases\n",
    "        so the algorithm can identify it 'ATGACCAGC'.\n",
    "\n",
    "    end : string\n",
    "        The algorithm will stop creating primers once it reaches that base.\n",
    "        For example, if you will stop mutating at the stop codon, copy a few more dna bases ie. 'TAAATGATT'.\n",
    "\n",
    "    output_file : str\n",
    "        File where the list of primers will be exported to. Only exports to excel.\n",
    "        Example: 'path/primers.xlsx'. \n",
    "\n",
    "    codon : str, default 'NNS'\n",
    "        Degenerate codon that will be used to create the primers. Check idt's website for a list\n",
    "        of all mixed bases and letter code (https://www.idtdna.com/pages/products/custom-dna-rna/mixed-bases).\n",
    "        This parameter should contain 3 letters, although can contain more.\n",
    "        \n",
    "    length_primer: int, default 15\n",
    "        Number of bases that the primers will have to each side of the mutated codon.\n",
    "        Total primer length will be 2*length_primer+3.\n",
    "\n",
    "    tm : int, default None\n",
    "        Melting temperature in Celsius of the primers. Will override length_primer.\n",
    "        If none, primers will have a total length of 2*length_primer+3\n",
    "\n",
    "    return_df : boolean, default False\n",
    "        If true, will export a dataframe with the primers.\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    df : pandas dataframe, optional\n",
    "        Dataframe containing the primers.\n",
    "\n",
    "    '''\n",
    "    # Transform to upper case\n",
    "    dna = dna.upper()\n",
    "\n",
    "    # Find the first and last codons\n",
    "    start_codon = dna.find(start.upper())\n",
    "    end_codon = dna.find(end.upper())\n",
    "\n",
    "    # loop through DNA and make a list with fp and second list with rp\n",
    "\n",
    "    label_fp = ['fp '+str(i) for i in range(0, int((end_codon-start_codon)/3))]\n",
    "    label_rp = ['rp '+str(i) for i in range(0, int((end_codon-start_codon)/3))]\n",
    "    forward_primers, reverse_primers = _create_primers_list(\n",
    "        dna, start_codon, end_codon, codon, length_primer, tm)\n",
    "\n",
    "    # Create dataframe\n",
    "    dictionary = {'FP_label': label_fp, 'FP_seq': forward_primers,\n",
    "                  'RP_label': label_rp, 'RP_seq': reverse_primers}\n",
    "    df = pd.DataFrame(dictionary)\n",
    "\n",
    "    # Export dataframe\n",
    "    if output_file:\n",
    "        df.to_excel(Path(output_file), sheet_name='Primers', index=False)\n",
    "\n",
    "    # Return dataframe\n",
    "    if return_df:\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-10-25T04:40:50.342336Z",
     "iopub.status.busy": "2020-10-25T04:40:50.342087Z",
     "iopub.status.idle": "2020-10-25T04:40:50.353207Z",
     "shell.execute_reply": "2020-10-25T04:40:50.352396Z",
     "shell.execute_reply.started": "2020-10-25T04:40:50.342311Z"
    }
   },
   "outputs": [],
   "source": [
    "def _create_primers_list(dna, start_codon, end_codon, codon, length_primer, tm):\n",
    "    '''Aux function to create list with fp and list with rp'''\n",
    "    forward_primers = []\n",
    "    reverse_primers = []\n",
    "    for codonposition in range(start_codon, end_codon, 3):\n",
    "        # Create fp, rp for that position\n",
    "        fp, rp = _primerdesign(dna, codon, codonposition, length_primer, tm)\n",
    "        # Append to list\n",
    "        forward_primers.append(fp)\n",
    "        reverse_primers.append(rp)\n",
    "    return forward_primers, reverse_primers\n",
    "\n",
    "\n",
    "def _reverse_complement(dna):\n",
    "    '''aux function that uses biopython to calculate the reverse complement of a DNA string.\n",
    "    Includes mixed-base code. More info in https://biopython.org/docs/1.75/api/Bio.Seq.html'''\n",
    "\n",
    "    # Needs to be converted to str\n",
    "    reverse_dna = str(Seq(dna, IUPAC.ambiguous_dna).reverse_complement())\n",
    "\n",
    "    return reverse_dna\n",
    "\n",
    "\n",
    "def _primerdesign(dna, codon, codonposition, length_primer, tm):\n",
    "    '''aux function to design the degenerate primers given a sequence and a codon position. \n",
    "    The length of the primer is fixed.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dna : string\n",
    "        DNA sequence containing the protein of study. \n",
    "        The DNA sequence should also contain at least 15 base pairs before the \n",
    "        starting ATG and 15 base pairs after the stop codon.\n",
    "\n",
    "    codon : str\n",
    "        Degenerate codon that will be used to create the primers. Check idt's website for a list\n",
    "        of all mixed bases and letter code (https://www.idtdna.com/pages/products/custom-dna-rna/mixed-bases).\n",
    "\n",
    "    codonposition : int\n",
    "        Position of the codon  to mutate with respect to the gene.\n",
    "\n",
    "    length_primer : int\n",
    "        Number of bases that the primers will have to each side of the mutated codon.\n",
    "\n",
    "    tm : int\n",
    "        melting temperature that the primers need to have\n",
    "\n",
    "    Returns\n",
    "    ---------\n",
    "    forward_primer, reverse_primer\n",
    "    '''\n",
    "\n",
    "    if tm:\n",
    "        # loop until tm is achieved\n",
    "        x = meltingT_fp = 6\n",
    "        while meltingT_fp < tm:\n",
    "            forward_primer = dna[(codonposition-x):codonposition] + \\\n",
    "                codon + dna[(codonposition+3):(codonposition+x)]\n",
    "            meltingT_fp = mt.Tm_NN(forward_primer)\n",
    "            x += 1\n",
    "    else:\n",
    "        forward_primer = dna[(codonposition-length_primer):codonposition] + \\\n",
    "            codon + dna[(codonposition+3):(codonposition+length_primer+3)]\n",
    "\n",
    "    reverse_primer = _reverse_complement(forward_primer)\n",
    "    return forward_primer, reverse_primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-10-25T04:40:28.898021Z",
     "iopub.status.busy": "2020-10-25T04:40:28.897817Z",
     "iopub.status.idle": "2020-10-25T04:40:29.066011Z",
     "shell.execute_reply": "2020-10-25T04:40:29.065318Z",
     "shell.execute_reply.started": "2020-10-25T04:40:28.898002Z"
    }
   },
   "outputs": [],
   "source": [
    "'''dna = 'TGTACAGTAATACAAGGGGTGTTATGGAAAAAATTATGCCGGAAGAAGAATACAGCGAATTTAAAGAACTGATTCTGCAGAAGGAACTGCACGTGGTGTATGCACTGAGCCACGTGTGTGGCCAGGATCGTACCCTGCTGGCCAGTATCTTACTGCGCATCTTTCTGCACGAGAAGCTGGAGAGCCTGTTACTGTGCACACTGAACGATCGCGAGATCAGCATGGAAGATGAAGCCACCACCCTGTTCCGCGCAACAACCCTGGCCAGCACCCTGATGGAGCAGTATATGAAAGCCACCGCCACCCAGTTCGTGCATCATGCCCTGAAAGATAGCATTTTAAAAATTATGGAAAGCAAACAGAGCTGCGAACTGAGCCCGAGCAAGCTGGAGAAAAACGAGGACGTGAACACCAACCTGACCCACCTGCTGAACATTCTGAGCGAACTGGTGGAAAAAATCTTTATGGCAAGCGAAATCCTGCCTCCGACCCTGCGTTACATCTACGGCTGCCTGCAGAAGAGCGTGCAGCATAAATGGCCGACCAATACCACCATGCGCACACGTGTGGTGAGCGGTTTTGTGTTCCTGCGTCTGATCTGCCCGGCAATCCTGAACCCGCGCATGTTCAACATCATTAGCGACAGCCCGAGTCCTATCGCAGCACGTACCCTGATCCTGGTGGCAAAAAGCGTGCAAAATCTGGCCAACCTGGTGGAATTTGGCGCCAAAGAGCCGTACATGGAAGGCGTGAATCCGTTTATCAAAAGTAACAAACATCGCATGATCATGTTCCTGGACGAACTGGGCAACGTTCCGGAACTGCCGGATACAACCGAACATAGTCGCACAGACCTGAGTCGTGACCTGGCCGCCCTGCATGAAATCTGCGTGGCCCATAGCGATGAGCTGCGCACACTGAGCAACGAGCGTGGCGCCCAGCAGCACGTGCTGAAGAAACTGCTGGCCATTACCGAACTGCTGCAACAAAAGCAGAACCAGTACACCAAAACCAACGACGTGCGTtatccgtatgatgtgccggattatgcgTAAccatcacttggctagaggcatc'\n",
    "start = ('ATGGAAAAAATTATGCCGGAAGAA')\n",
    "end = ('tatccgtatgatgtgccggattatgcgTA')\n",
    "\n",
    "df_prim = generate_primers(dna, start, end, output_file=None,\n",
    "                           codon='NNS', length_primer=15, tm=60, return_df=True)'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''def _enumerate_dnavariants(wtSeqList, codon_list, dna_sequence):\n",
    "\n",
    "    # First instance that we see the wt?\n",
    "    firstwtseq = False\n",
    "\n",
    "    # Loop over codons\n",
    "    for position in range(0,len(wtSeqList)):\n",
    "        for codons in (codon_list):\n",
    "            variant = ''.join(wtSeqList[0:position]) + \\\n",
    "                codons + ''.join(wtSeqList[position+1:])\n",
    "            if (variant == dna_sequence):  # Store redundant wild-types\n",
    "                if firstwtseq:\n",
    "                    variant = 'wtSeq' + str(position)\n",
    "                firstwtseq = True\n",
    "            variants[variant] = 0\n",
    "    return variants'''\n",
    "\n",
    "# options to return all mutants, or filter out synonymous wt alleles"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
