{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-12T22:36:10.914292Z",
     "start_time": "2020-10-12T22:36:06.031777Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-20T20:32:55.835642Z",
     "iopub.status.busy": "2020-10-20T20:32:55.835273Z",
     "iopub.status.idle": "2020-10-20T20:32:57.635005Z",
     "shell.execute_reply": "2020-10-20T20:32:57.633914Z",
     "shell.execute_reply.started": "2020-10-20T20:32:55.835598Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "importing Jupyter notebook from code_3D.ipynb\n",
      "importing Jupyter notebook from code_heatmaps.ipynb\n",
      "importing Jupyter notebook from code_kernel.ipynb\n",
      "importing Jupyter notebook from code_other.ipynb\n",
      "importing Jupyter notebook from code_plotly.ipynb\n"
     ]
    }
   ],
   "source": [
    "# Regular libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools\n",
    "from collections import defaultdict\n",
    "from Bio.Seq import Seq\n",
    "from pathlib import Path\n",
    "from typing import Union"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Internal Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.620761Z",
     "start_time": "2020-10-06T23:17:05.598156Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-20T20:32:59.927935Z",
     "iopub.status.busy": "2020-10-20T20:32:59.927683Z",
     "iopub.status.idle": "2020-10-20T20:32:59.949994Z",
     "shell.execute_reply": "2020-10-20T20:32:59.948592Z",
     "shell.execute_reply.started": "2020-10-20T20:32:59.927907Z"
    }
   },
   "outputs": [],
   "source": [
    "def _transform_dataset(dataset, sequence, aminoacids, start_position, fillna):\n",
    "    '''\n",
    "    Internal function that constructs a dataframe from user inputs\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dataset, sequence, aminoacids, start_position,fillna\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    Dataframe containing [Position, Sequence, Aminoacid, Variant, Score]\n",
    "    '''\n",
    "\n",
    "    # make a dataframe\n",
    "    df = pd.DataFrame()\n",
    "\n",
    "    # Define Columns\n",
    "    df['Sequence'] = np.ravel([[aa]*len(aminoacids) for aa in sequence])\n",
    "\n",
    "    # Create column with position label\n",
    "    df['Position'] = np.ravel(\n",
    "        [[i]*len(aminoacids) for i in range(start_position, len(dataset[0])+start_position)])\n",
    "    df['Aminoacid'] = aminoacids * len(dataset[0])\n",
    "    df['Variant'] = df['Sequence']+df['Position'].astype(str)+df['Aminoacid']\n",
    "    df['Score'] = np.ravel(dataset.T)\n",
    "    df['Score_NaN'] = np.ravel(dataset.T)\n",
    "\n",
    "    # Eliminate NaNs\n",
    "    df['Score'].fillna(fillna, inplace=True)\n",
    "\n",
    "    # Eliminate stop codons\n",
    "    df_clean = df[df['Aminoacid'] != '*'].copy()\n",
    "\n",
    "    return df, df_clean\n",
    "\n",
    "\n",
    "def _transform_sequence(dataset, sequence, start_position):\n",
    "    '''\n",
    "    Internal function that trims the input sequence\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dataset, sequence, start_position\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    string containing trimmed sequence\n",
    "    '''\n",
    "\n",
    "    # truncate sequence\n",
    "    trimmedsequence = sequence[start_position -\n",
    "                               1:len(dataset[0])+start_position-1]\n",
    "\n",
    "    return trimmedsequence\n",
    "\n",
    "\n",
    "def _transform_secondary(dataset, secondary, start_position, aminoacids):\n",
    "    '''\n",
    "    Internal function that trims the input secondary structure\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    dataset, sequence, start_position\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    list containing trimmed secondary structure (20 times each element)\n",
    "    '''\n",
    "\n",
    "    # Convert lists of lists to list\n",
    "    secondary_list = list(itertools.chain.from_iterable(secondary))\n",
    "\n",
    "    # Truncate list\n",
    "    trimmedsecondary = secondary_list[start_position -\n",
    "                                      1:len(dataset[0])+start_position-1]\n",
    "\n",
    "    # Multiply each element by number of aminoacids. not use stop codon\n",
    "    aminoacids = list(np.copy(aminoacids))\n",
    "    if '*' in aminoacids:\n",
    "        aminoacids.remove('*')\n",
    "    secondary_dup = [x for item in trimmedsecondary for x in itertools.repeat(\n",
    "        item, len(aminoacids))]\n",
    "\n",
    "    return trimmedsecondary, secondary_dup\n",
    "\n",
    "\n",
    "def _convert_to_df(dataset, sequence, aminoacids, startposition):\n",
    "    '''\n",
    "    Convertds np.array with stored enrichment scores into a dataframe\n",
    "    Makes a copy of data\n",
    "\n",
    "    Returns dataframe\n",
    "\n",
    "    '''\n",
    "    df = pd.DataFrame()\n",
    "    df['Aminoacid'] = list(aminoacids) * len(dataset[0])\n",
    "    df['Position'] = np.ravel(\n",
    "        [[i]*len(aminoacids) for i in range(startposition, len(dataset[0])+startposition)])\n",
    "    df['Sequence'] = np.ravel([[i]*len(aminoacids)\n",
    "                               for i in sequence[:len(dataset[0])]])\n",
    "    df['Score'] = np.copy(dataset.T).ravel()\n",
    "    return df\n",
    "\n",
    "\n",
    "def _df_rearrange(df, new_order, values='Score',show_snv = False):\n",
    "    '''\n",
    "    convert a df into a numpy array for mutagenesis data. \n",
    "    Allows the option of keeping NaN scores\n",
    "\n",
    "    Returns copy\n",
    "    '''\n",
    "    dfcopy = df.copy()\n",
    "    \n",
    "    # If only SNVs, turn rest to NaN\n",
    "    if show_snv is True:\n",
    "        dfcopy.loc[dfcopy['SNV?']==False, values] = np.nan\n",
    "    \n",
    "    df_pivoted = dfcopy.pivot_table(values=values, index='Aminoacid',\n",
    "                                    columns=['Position'], dropna=False)\n",
    "    df_reindexed = df_pivoted.reindex(index=list(new_order))\n",
    "\n",
    "    return df_reindexed\n",
    "\n",
    "\n",
    "def _common(a, b):\n",
    "    '''\n",
    "    return common elements of two lists\n",
    "    '''\n",
    "    c = [value for value in a if value in b]\n",
    "    return c\n",
    "\n",
    "\n",
    "def _transpose(df, values='Score'):\n",
    "    '''\n",
    "    convert a df into a numpy array for mutagenesis data\n",
    "\n",
    "    Returns copy\n",
    "    '''\n",
    "    df = df.pivot_table(values=values, index='Aminoacid',\n",
    "                        columns=['Position']).T\n",
    "    return df\n",
    "\n",
    "\n",
    "def _select_aa(df, selection, values='Score'):\n",
    "    '''returns copy'''\n",
    "    df = _transpose(df.copy(), values)\n",
    "\n",
    "    df = df[selection].T\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def _savefile(fig, temp_kwargs):\n",
    "    '''DEPRECATED\n",
    "    Save file function'''\n",
    "    if temp_kwargs['savefile'] is True:\n",
    "        filename = temp_kwargs['outputfilepath'] + \\\n",
    "            temp_kwargs['outputfilename']+\".\"+temp_kwargs['outputformat']\n",
    "        fig.savefig(filename, format=temp_kwargs['outputformat'],\n",
    "                    bbox_inches='tight', dpi=temp_kwargs['dpi'], transparent=True)\n",
    "    return\n",
    "\n",
    "def _save_work(fig, output_file, temp_kwargs):\n",
    "    '''Save file function using pathlib'''\n",
    "    if output_file:\n",
    "        fig.savefig(Path(output_file), format=Path(output_file).suffix.strip('.'),\n",
    "                    bbox_inches='tight', dpi=temp_kwargs['dpi'], transparent=True)\n",
    "    return\n",
    "\n",
    "\n",
    "def parse_pivot(df_imported, col_variant = 'variant', col_data = 'DMS',\n",
    "               fill_value = np.nan):\n",
    "    '''\n",
    "    Parses a dataframe that contains saturation mutagenesis data in the Variant/Scores format.\n",
    "    \n",
    "    Parameters\n",
    "    -----------\n",
    "    df_imported : pandas dataframe\n",
    "        Dataframe with the data imported with pd.read_excel.\n",
    "    \n",
    "    col_variant : str, default 'variant'\n",
    "        Name of the column that contains the variants (ie T31A).\n",
    "        \n",
    "    col_data : str, default 'DMS'\n",
    "        Name of the column that contains the saturation mutagenesis scores.\n",
    "    \n",
    "    fill_value : float, default np.nan\n",
    "        What number to replace values that are omitted. It is possible that your \n",
    "        dataset does not have a wt value.\n",
    "        \n",
    "    Returns\n",
    "    --------\n",
    "    df_pivoted : pandas dataframe\n",
    "        Dataframe that has been pivoted. Values are the saturation mutagenesis data. Columns are \n",
    "        the amino acid substitutions. Rows are the positions of the protein.\n",
    "    \n",
    "    sequence : list\n",
    "        List of the amino acids that form the protein sequence.\n",
    "    '''\n",
    "    \n",
    "    # Copy\n",
    "    df = df_imported.copy()\n",
    "    \n",
    "    # Extract position and amino acids that are being mutated\n",
    "    df['Position'] = df[col_variant].str.extract('(\\d+)').astype(int)\n",
    "    df['Original'] = df[col_variant].str[0:1]\n",
    "    df['Substitution'] = df[col_variant].str[-1:]\n",
    "    \n",
    "    # Get sequence\n",
    "    sequence = list(df.groupby(by=['Position', 'Original'], as_index=False, group_keys=False).sum() ['Original'])\n",
    "\n",
    "    # Pivot\n",
    "    df_pivoted = df.pivot_table(index='Substitution',columns = 'Position', \n",
    "                                values=col_data, fill_value = fill_value, dropna=False)\n",
    "    \n",
    "    return df_pivoted, sequence\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNV internal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _select_nonSNV(df):\n",
    "    '''\n",
    "    Generate a dataframe that contains the non-SNV variants and the enrichment score\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    df : pd.dataframe\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    Dataframe containing a column of variants that are non-SNV, and the Score.\n",
    "    '''\n",
    "    # Dataframe with SNV\n",
    "    SNV = _select_SNV(df)\n",
    "\n",
    "    # Merge and eliminate duplicates. Keep Non-SNV\n",
    "    NonSNV = pd.concat([SNV, df], sort=False)[['Position', 'Variant', 'Score','Score_NaN']]\n",
    "    NonSNV.drop_duplicates(subset='Variant', keep=False, inplace=True)\n",
    "\n",
    "    return NonSNV\n",
    "\n",
    "\n",
    "def _select_SNV(df):\n",
    "    '''\n",
    "    Select for SNV variants in DSM dataset\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    df : pandas dataframe containing DSM data\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    Modified dataframe('Variant','Score') where 'SNV?'== True. Returns copy\n",
    "    '''\n",
    "\n",
    "    # Use _add_SNV_boolean funciton\n",
    "    df = _add_SNV_boolean(df.copy())\n",
    "\n",
    "    # Select SNV? == True only\n",
    "    df = df[df['SNV?'] == True].copy()\n",
    "\n",
    "    # Select columns of interest\n",
    "    df = df[['Position', 'Variant', 'Score','Score_NaN']].copy()\n",
    "\n",
    "    # Reset index\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def _aminoacids_snv(aa1, aa2, codontable):\n",
    "    '''\n",
    "    Determine if two amino acids are snv (one base difference)\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    aa1 : str\n",
    "    aa2 : str\n",
    "    codontable : dict (did not want to generate each time I run the function)\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    boolean, True/False\n",
    "    '''\n",
    "    # Convert amino acids to codons\n",
    "    codons1 = codontable[aa1]\n",
    "    codons2 = codontable[aa2]\n",
    "\n",
    "    # Generate a list of combination pairs between all codons in aa1 and aa2\n",
    "    codon_combinations = list(itertools.product(codons1, codons2))\n",
    "\n",
    "    # If one pair of combinations is a SNV, then return True\n",
    "    for combination in codon_combinations:\n",
    "        if _codons_pointmutants(combination[0], combination[1]) == True:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "\n",
    "def _add_SNV_boolean(df):\n",
    "    '''\n",
    "    Add a column to dataframe indication if the variant is a SNV or not\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    df : pandas dataframe containing DSM data\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    Modified dataframe. Returns copy\n",
    "    '''\n",
    "\n",
    "    # Generate dictionary with aa and codon translation\n",
    "    codontable = _dict_codontoaa()\n",
    "\n",
    "    # Add column with True/False input\n",
    "    df['SNV?'] = df.apply(lambda x: _aminoacids_snv(\n",
    "        x['Sequence'], x['Aminoacid'], codontable), axis=1)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def _codons_pointmutants(codon1, codon2):\n",
    "    '''\n",
    "    Determine if two codons are SNV. Returns a boolean\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    codon1 : str\n",
    "    codon2 : str\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    boolean, True/False\n",
    "    '''\n",
    "    counter_occurrences = 0\n",
    "    for index, base1 in enumerate(codon1):\n",
    "        base2 = list(codon2)[index]\n",
    "        if base1 == base2:\n",
    "            counter_occurrences = counter_occurrences+1\n",
    "    if counter_occurrences > 1:\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "\n",
    "def _are_pointmutants(aa, seqbase):\n",
    "    '''\n",
    "    converts the amino acid to all possible degenerate codons and then checks if they are point mutants\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    aa: str\n",
    "    seqbase: str    \n",
    "\n",
    "    Returns \n",
    "    --------\n",
    "    Boolean\n",
    "    '''\n",
    "    codontoaadict = _dict_codontoaa()\n",
    "    pointmutants = False\n",
    "    for codon in _codontoaadict[aa]:\n",
    "        if _codons_pointmutants(seqbase, codon):\n",
    "            pointmutants = True\n",
    "    return pointmutants\n",
    "\n",
    "\n",
    "def _are_pointmutants_list(aa, seqbase_list):\n",
    "    '''\n",
    "    converts the amino acid to all possible degenerate codons and then checks if they are point mutants\n",
    "    Same as _are_pointmutants but in list format\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    aa: str\n",
    "    seqbase_list: list of str    \n",
    "\n",
    "    Returns \n",
    "    --------\n",
    "    List of Boolean\n",
    "    '''\n",
    "    pointmutants_list = []\n",
    "\n",
    "    for seqbase in seqbase_list:\n",
    "        pointmutants_list.append(_are_pointmutants(aa, seqbase))\n",
    "    return pointmutants_list\n",
    "\n",
    "\n",
    "def _dict_codontoaa():\n",
    "    '''\n",
    "    Generates a dictionary with all amino acids and all possible codons.\n",
    "    aa is the aminoacid of the mutation and seqbase is the original codon of the wtsequence\n",
    "    '''\n",
    "    bases = ['T', 'C', 'A', 'G']\n",
    "    codons = [a+b+c for a in bases for b in bases for c in bases]\n",
    "    aminoacids = list(\n",
    "        'FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG')\n",
    "\n",
    "    # dictionary with more than one value for each key\n",
    "    codontoaadict = defaultdict(list)\n",
    "    for codon, aminoacid in zip(codons, aminoacids):\n",
    "        codontoaadict[aminoacid].append(codon)\n",
    "    return codontoaadict\n",
    "\n",
    "\n",
    "def _aatocodons(aminoacid):\n",
    "    '''\n",
    "    Inputs an aminoacid, returns all codons. Used dict_codontoaa()\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    aminoacid : str\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    List with all the codons that code for that amino acid\n",
    "    '''\n",
    "\n",
    "    # Dictionary with all codons and aa\n",
    "    codontoaadict = _dict_codontoaa()\n",
    "\n",
    "    # Codons for that amino acid\n",
    "    codons = codontoaadict[aminoacid]\n",
    "\n",
    "    return codons\n",
    "\n",
    "\n",
    "def _aatocodons_df(df, namecolumn):\n",
    "    '''\n",
    "    Inputs a dataframe with a column of amino acids, returns all syn for each amino acidcodons. \n",
    "    Used dict_codontoaa() and _aatocodons\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "    namecolumn : str\n",
    "        name of the column containing the amino acids\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    dataframe with a column containing all the codons that code for that amino acid. Returns copy\n",
    "    '''\n",
    "    # Copy df\n",
    "    df = df.copy()\n",
    "\n",
    "    # Calculate each possible codon for every amino acid\n",
    "    df['Codons_' +\n",
    "        namecolumn] = df.apply(lambda x: _aatocodons(x[namecolumn]), axis=1)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scatter Internal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _process_bypointmutant(self, obj):\n",
    "    # truncate so both datasets have same length and delete stop codons\n",
    "    minlength = min(len(self.dataframe), len(obj.dataframe))\n",
    "    df = pd.DataFrame()\n",
    "    df['dataset_1'] = list(self.dataframe['Score_NaN'])[:minlength]\n",
    "    df['dataset_2'] = list(obj.dataframe['Score_NaN'])[:minlength]\n",
    "    df['Variant'] = list(self.dataframe['Variant'])[:minlength]\n",
    "    \n",
    "    # eliminate Nans\n",
    "    df.dropna(how='any', inplace=True)\n",
    "    return df\n",
    "\n",
    "\n",
    "def _process_meanresidue(self, obj):\n",
    "    # truncate so both datasets have same length and delete stop codons\n",
    "    dataset_1 = self.dataframe.groupby(['Position'], as_index=False).mean()\n",
    "    dataset_2 = obj.dataframe.groupby(['Position'], as_index=False).mean()\n",
    "    minlength = min(len(dataset_1), len(dataset_2))\n",
    "\n",
    "    # convert to dataframe and eliminate Nans\n",
    "    df = pd.DataFrame()\n",
    "    df['dataset_1'] = list(dataset_1['Score'])[0:minlength]\n",
    "    df['dataset_2'] = list(dataset_2['Score'])[0:minlength]\n",
    "    df['Position'] = list(dataset_1['Position'])[0:minlength]\n",
    "    df['d1 - d2'] = df['dataset_1'] - df['dataset_2']\n",
    "    df.dropna(how='any', inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _color_data(row, color_gof, color_lof):\n",
    "    if row['Score'] > 0:\n",
    "        return color_gof\n",
    "    else:\n",
    "        return color_lof"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# To manipulate reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-10-27T03:35:01.307076Z",
     "iopub.status.busy": "2020-10-27T03:35:01.306728Z",
     "iopub.status.idle": "2020-10-27T03:35:01.316873Z",
     "shell.execute_reply": "2020-10-27T03:35:01.315943Z",
     "shell.execute_reply.started": "2020-10-27T03:35:01.307020Z"
    }
   },
   "outputs": [],
   "source": [
    "def _translate_codons(df):\n",
    "    '''Translate the index of the df from codons to AA'''\n",
    "    list_aa = [str(Seq(codon).translate()) for codon in list(df.index)]\n",
    "    return list_aa\n",
    "\n",
    "\n",
    "def _is_DNA(df):\n",
    "    '''Check if the index of the dataframe are the DNA codons'''\n",
    "    aminoacids = 'DEFHIKLMNPQRSVWY*'\n",
    "    for aa in aminoacids:\n",
    "        if aa in ''.join(list(df.index)):\n",
    "            return False\n",
    "        return True"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "100px",
    "left": "21px",
    "top": "69.2px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.4,
   "position": {
    "height": "423.4px",
    "left": "944.6px",
    "right": "20px",
    "top": "119px",
    "width": "315.4px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "block",
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
