{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-12T22:36:10.914292Z",
     "start_time": "2020-10-12T22:36:06.031777Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools\n",
    "import copy\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.ticker as ticker\n",
    "from pathlib import Path\n",
    "from typing import Union\n",
    "\n",
    "# kwargs and parameters\n",
    "try:\n",
    "    import import_notebook\n",
    "except ModuleNotFoundError:\n",
    "    pass\n",
    "import code_kwargs\n",
    "import code_utils\n",
    "import code_heatmaps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mean Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bar graph Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:04.971508Z",
     "start_time": "2020-10-06T23:17:04.950334Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_mean(self, mode='mean', show_cartoon=False,\n",
    "              output_file: Union[None, str, Path] = None, **kwargs):\n",
    "    '''\n",
    "    Plot in a bargraph the mean enrichment for each residue of the protein. Red for gain of function, blue for loss of function\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "\n",
    "    mode : str, default 'mean'\n",
    "        Specify what enrichment scores to show. If mode = 'mean', it will show the mean of \n",
    "        each position. If mode = 'A', it will show the alanine substitution profile. Can be \n",
    "        used for each amino acid. Use the one-letter code and upper case.\n",
    "\n",
    "    show_carton : boolean, default False\n",
    "        If true, the plot will display a cartoon with the secondary structure. The user must have added the secondary structure to the object. \n",
    "\n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "\n",
    "    **kwargs : other keyword arguments\n",
    "        color_gof : str, default 'red'\n",
    "            Choose color to color positions with an enrichment score > 0.\n",
    "\n",
    "        color_lof : str, default 'blue'\n",
    "            Choose color to color positions with an enrichment score < 0.\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "    '''\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (3, 2.5))\n",
    "    temp_kwargs['y_label'] = kwargs.get('y_label', r'$∆E^i_x$')\n",
    "\n",
    "    # load parameters\n",
    "    _parameters_mean()\n",
    "\n",
    "    # Select grouping\n",
    "    if mode == 'mean':\n",
    "        df = self.dataframe.groupby('Position', as_index=False).mean()\n",
    "    else:\n",
    "        df = self.dataframe.loc[self.dataframe['Aminoacid'] == mode].copy()\n",
    "\n",
    "    df['Color'] = df.apply(code_utils._color_data, axis=1, args=(temp_kwargs['color_gof'],\n",
    "                                                                 temp_kwargs['color_lof']))\n",
    "\n",
    "    # make figure\n",
    "    if show_cartoon:\n",
    "        fig = plt.figure(figsize=temp_kwargs['figsize'])\n",
    "        gs = gridspec.GridSpec(nrows=2, ncols=1, height_ratios=[5, 1])\n",
    "        ax = plt.subplot(gs[0])\n",
    "    else:\n",
    "        fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    width = 1.2\n",
    "\n",
    "    # Color based on values\n",
    "    ax.bar(df['Position'], df['Score'], width, color=df['Color'], snap=False)\n",
    "\n",
    "    # axes parameters\n",
    "    ax.set_ylim(temp_kwargs['yscale'])\n",
    "    ax.set_ylabel(temp_kwargs['y_label'], fontsize=10,\n",
    "                  fontname=\"Arial\", color='k', labelpad=10, rotation=0)\n",
    "    ax.set_xticks(np.arange(self.start_position,\n",
    "                            len(df)+self.start_position, 20))\n",
    "    ax.set_xlabel('Residue', fontsize=10,\n",
    "                  fontname=\"Arial\", color='k', labelpad=4)\n",
    "    ax.set_xlim(self.start_position-0.1, len(df)+self.start_position-1+0.1)\n",
    "\n",
    "    # cartoon\n",
    "    title_pad = 0\n",
    "    if show_cartoon:\n",
    "        code_heatmaps._generate_cartoon(self, gs, 1, temp_kwargs['cartoon_colors'],\n",
    "                                        bottom_space=-0.78, show_labels=False)\n",
    "        title_pad = 2.5\n",
    "\n",
    "    # Plot title\n",
    "    plt.title(temp_kwargs['title'], fontsize=12, fontname='Arial', color='k',\n",
    "              pad=title_pad)\n",
    "\n",
    "    # Put text labels\n",
    "    _inputtext(temp_kwargs['text_labels'])\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return\n",
    "\n",
    "\n",
    "def _parameters_mean():\n",
    "    # normal font\n",
    "    rcParams['font.family'] = 'sans-serif'\n",
    "    rcParams['font.sans-serif'] = ['Arial']\n",
    "\n",
    "    # math font\n",
    "    rcParams['mathtext.fontset'] = 'custom'\n",
    "    rcParams['mathtext.rm'] = 'Arial'\n",
    "    rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "    # Parameters for all graphs\n",
    "    rcParams['xtick.labelsize'] = 9\n",
    "    rcParams['ytick.labelsize'] = 9\n",
    "\n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare two proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:04.997893Z",
     "start_time": "2020-10-06T23:17:04.978050Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_meandifferential(self, obj2, show_cartoon=False,\n",
    "                          output_file: Union[None, str, Path] = None, **kwargs):\n",
    "    '''\n",
    "    Plot the mean positional difference between two experiments\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "\n",
    "    obj2 : another Screen object to compare with \n",
    "\n",
    "    show_carton : boolean, default False\n",
    "        If true, the plot will display a cartoon with the secondary structure. The user must have added the secondary structure to the object. \n",
    "\n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "\n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "    '''\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (3, 2.5))\n",
    "    temp_kwargs['yscale'] = kwargs.get('yscale', (-1, 1))\n",
    "    temp_kwargs['y_label'] = kwargs.get(\n",
    "        'y_label', r'Mean Differential $∆E^i_x$')\n",
    "\n",
    "    # load parameters\n",
    "    _parameters_mean()\n",
    "\n",
    "    # make pandas\n",
    "    df = code_utils._process_meanresidue(self, obj2)\n",
    "\n",
    "    # make cartoon\n",
    "    if show_cartoon:\n",
    "        fig = plt.figure(figsize=temp_kwargs['figsize'])\n",
    "        gs = gridspec.GridSpec(nrows=2, ncols=1, height_ratios=[5, 1])\n",
    "        ax = plt.subplot(gs[0])\n",
    "    else:\n",
    "        fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "\n",
    "    # plot\n",
    "    ax.plot(df['Position'], df['d1 - d2'], color='k')\n",
    "\n",
    "    # axes parameters\n",
    "    ax.set_ylim(temp_kwargs['yscale'])\n",
    "    ax.set_ylabel(temp_kwargs['y_label'], fontsize=10, fontname=\"Arial\",\n",
    "                  color='k', labelpad=-5, rotation=90)\n",
    "    ax.set_xticks(np.arange(self.start_position,\n",
    "                            len(df)+self.start_position, 20))\n",
    "    ax.set_xlabel('Residue', fontsize=10,\n",
    "                  fontname=\"Arial\", color='k', labelpad=4)\n",
    "    ax.set_xlim(self.start_position-0.1, len(df)+self.start_position-1+0.1)\n",
    "\n",
    "    # cartoon\n",
    "    title_pad = 0\n",
    "    if show_cartoon:\n",
    "        title_pad = 2.5\n",
    "        obj = obj2\n",
    "        if len(self.dataframe) < len(obj2.dataframe):\n",
    "            obj = self\n",
    "        code_heatmaps._generate_cartoon(obj, gs, 1, temp_kwargs['cartoon_colors'],\n",
    "                                        bottom_space=-0.78, show_labels=False)\n",
    "\n",
    "    # title\n",
    "    ax.set_title(temp_kwargs['title'], fontsize=12, fontname='Arial',\n",
    "                 color='k', pad=title_pad)\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bar graph Counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.015657Z",
     "start_time": "2020-10-06T23:17:05.001297Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_meancounts(self, output_file: Union[None, str, Path] = None, **kwargs):\n",
    "    '''\n",
    "    Plot in a bargraph the mean counts for each residue of the protein.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "\n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "    '''\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (10, 4))\n",
    "    temp_kwargs['y_label'] = kwargs.get('y_label', 'Mean counts')\n",
    "\n",
    "    # load parameters\n",
    "    _parameters_mean()\n",
    "\n",
    "    # make figure\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    ax = self.dataframe.mean().T.plot.bar(ax=ax, color='red')\n",
    "\n",
    "    # axes parameters\n",
    "    ax.set_ylim(temp_kwargs['yscale'])\n",
    "    ax.set_ylabel(temp_kwargs['y_label'], fontsize=12,\n",
    "                  fontname=\"Arial\", color='k', labelpad=0, rotation=90)\n",
    "    ax.set_xlabel('Amino acid position', fontsize=12,\n",
    "                  fontname=\"Arial\", color='k', labelpad=4)\n",
    "    ax.set_xticklabels(self.positions)\n",
    "\n",
    "    # Title\n",
    "    plt.title(temp_kwargs['title'], fontsize=14, fontname='Arial',\n",
    "              color='k', pad=0)\n",
    "\n",
    "    # Put text labels\n",
    "    _inputtext(temp_kwargs['text_labels'])\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return\n",
    "\n",
    "\n",
    "def _inputtext(text_entries):\n",
    "    '''the user can input text as a variable by manually giving the coordinates'''\n",
    "    if text_entries:\n",
    "        for entry in text_entries:\n",
    "            plt.text(entry[0], entry[1], entry[2])\n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Positional"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.032278Z",
     "start_time": "2020-10-06T23:17:05.018682Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_position(self, position, output_file: Union[None, str, Path] = None,\n",
    "                  **kwargs):\n",
    "    '''\n",
    "    Choose a position and plot in a bargraph the enrichment score for each substitution.\n",
    "    Red for gain of function, blue for loss of function.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    self : object from class *Screen*\n",
    "\n",
    "    position : int\n",
    "        number of residue of the protein to display.\n",
    "\n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "\n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "    '''\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (3.5, 2))\n",
    "    temp_kwargs['yscale'] = kwargs.get('yscale', (-1, 1))\n",
    "    temp_kwargs['y_label'] = kwargs.get('y_label', r'$∆E^i_x$')\n",
    "\n",
    "    # load parameters\n",
    "    _parameters_mean()\n",
    "\n",
    "    # Select position\n",
    "    df = self.dataframe.loc[self.dataframe['Position'] == position].copy()\n",
    "\n",
    "    # Color\n",
    "    df['Color'] = df.apply(code_utils._color_data, axis=1,\n",
    "                           args=(temp_kwargs['color_gof'], temp_kwargs['color_lof']))\n",
    "\n",
    "    # make figure\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    width = 0.5\n",
    "\n",
    "    # Color based on values\n",
    "    ax.bar(df['Aminoacid'], df['Score'], width, color=df['Color'], ec='k')\n",
    "\n",
    "    # axes parameters\n",
    "    ax.set_ylim(temp_kwargs['yscale'])\n",
    "    ax.set_ylabel(temp_kwargs['y_label'], fontsize=10,\n",
    "                  fontname=\"Arial\", color='k', labelpad=10, rotation=0)\n",
    "\n",
    "    ax.set_xlabel('Residue', fontsize=10,\n",
    "                  fontname=\"Arial\", color='k', labelpad=4)\n",
    "    plt.title(temp_kwargs['title'], fontsize=12, fontname='Arial', color='k')\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stacked bar cumulative %AA representation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_library_representation(self,output_file: Union[None, str, Path] = None,\n",
    "                                **kwargs):\n",
    "    '''\n",
    "     Generates a cumulative stacked bar plot. Each bar represents an amino acid position,\n",
    "     and each color indicates the observed variant frequency.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'.\n",
    "\n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "    '''\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (10, 4))\n",
    "\n",
    "    # load parameters\n",
    "    _parameters_mean()\n",
    "\n",
    "    # Transform data\n",
    "    df_grouped = _group_codons_to_aa(self)\n",
    "    df_percentage = _percentage_column(df_grouped)\n",
    "    \n",
    "    # colors\n",
    "    colors = plt.cm.tab20(np.linspace(0, 1, 20))\n",
    "\n",
    "    # plot\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    ax = df_percentage.T.plot.bar(stacked=True, ax=ax, color=colors)\n",
    "\n",
    "    # axes parameters\n",
    "    ax.set_ylim(0,100)\n",
    "    ax.set_ylabel('Cumulative % AA representation', fontsize=12,\n",
    "                  fontname=\"Arial\", color='k', labelpad=0, rotation=90)\n",
    "    ax.yaxis.set_major_formatter(ticker.PercentFormatter())\n",
    "\n",
    "    ax.set_xlabel('Amino acid position', fontsize=12,\n",
    "                  fontname=\"Arial\", color='k', labelpad=4)\n",
    "    ax.set_xticklabels(self.positions)\n",
    "\n",
    "\n",
    "    plt.title(temp_kwargs['title'], fontsize=14, fontname='Arial', color='k', pad=20)\n",
    "\n",
    "    # legend\n",
    "    plt.legend(markerscale=0.5, frameon=False, framealpha=0,\n",
    "               handlelength=0.75, handletextpad=0.25, ncol=len(df_percentage), \n",
    "               columnspacing=0.75,bbox_to_anchor=(0.5, 1.09), loc='upper center')\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _group_codons_to_aa(self):\n",
    "    '''Group different codons that are synonymous. Returns sum of counts'''\n",
    "    # copy df\n",
    "    df = self.dataframe.copy()\n",
    "\n",
    "    df['Aminoacid'] = self.aminoacids\n",
    "\n",
    "    # Group by mean\n",
    "    df = df.groupby(as_index=True, by='Aminoacid', sort=False).sum()\n",
    "    return df\n",
    "\n",
    "def _percentage_column(df):\n",
    "    '''Make the percentage per column'''\n",
    "    return df/df.sum(axis=0)*100"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "100px",
    "left": "21px",
    "top": "69.2px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.4,
   "position": {
    "height": "423.4px",
    "left": "944.6px",
    "right": "20px",
    "top": "119px",
    "width": "315.4px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "block",
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
