{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-12T22:36:10.914292Z",
     "start_time": "2020-10-12T22:36:06.031777Z"
    },
    "execution": {
     "iopub.status.idle": "2020-10-11T02:23:58.137436Z",
     "shell.execute_reply": "2020-10-11T02:23:58.136608Z",
     "shell.execute_reply.started": "2020-10-11T02:23:53.211168Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import itertools\n",
    "import copy\n",
    "from scipy import stats\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.patches as patches\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import matplotlib.ticker as ticker\n",
    "from sklearn import metrics\n",
    "from sklearn.decomposition import PCA\n",
    "from adjustText import adjust_text\n",
    "import freesasa\n",
    "from os import path\n",
    "import os\n",
    "import logomaker\n",
    "from pathlib import Path\n",
    "from typing import Union\n",
    "import plotly.express as px\n",
    "import plotly.io as pio\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "# kwargs and parameters\n",
    "try:\n",
    "    import import_notebook\n",
    "except ModuleNotFoundError:\n",
    "    pass\n",
    "import code_kwargs\n",
    "import code_utils\n",
    "import code_heatmaps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Heatmap correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.314894Z",
     "start_time": "2020-10-06T23:17:05.225096Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.543441Z",
     "iopub.status.busy": "2020-10-11T02:23:58.542705Z",
     "iopub.status.idle": "2020-10-11T02:23:58.560141Z",
     "shell.execute_reply": "2020-10-11T02:23:58.558878Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.543414Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_correlation(self, output_file: Union[None, str, Path] = None,\n",
    "                     **kwargs):\n",
    "    '''\n",
    "    Generate a correlation of each amino acid\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                \n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "    '''\n",
    "\n",
    "    # load font parameters\n",
    "    code_kwargs._font_parameters()\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "\n",
    "    # load labels\n",
    "    temp_kwargs['color_sequencelabels'] = code_heatmaps._labels(self.start_position)[0]\n",
    "    temp_kwargs['number_sequencelabels'] = code_heatmaps._labels(self.start_position)[1]\n",
    "\n",
    "    # calculate correlation heatmap\n",
    "    dataset = _calculate_correlation(\n",
    "        self.dataframe_stopcodons, temp_kwargs['neworder_aminoacids'])\n",
    "\n",
    "    # declare figure and subplots\n",
    "    coeff = len(dataset.columns)/19*1.05\n",
    "    fig = plt.figure(figsize=(2.5*coeff, 2.5))\n",
    "    gs = gridspec.GridSpec(nrows=1, ncols=1)\n",
    "    ax = plt.subplot(gs[0])\n",
    "\n",
    "    # main heatmap\n",
    "    heatmap = ax.pcolor(dataset.corr(), vmin=temp_kwargs['colorbar_scale'][0], vmax=temp_kwargs['colorbar_scale'][1],\n",
    "                        cmap='Greys', edgecolors='k', linewidths=0.2, color='darkgrey')\n",
    "\n",
    "    # ____________axes manipulation____________________________________________\n",
    "    # put the major ticks at the middle of each cell\n",
    "    ax.set_xticks(np.arange(dataset.shape[1]) + 0.5, minor=False)\n",
    "    ax.set_yticks(np.arange(dataset.shape[0]) + 0.5, minor=False)\n",
    "\n",
    "    # position of axis labels\n",
    "    ax.tick_params('x', direction='out', pad=-2.5)\n",
    "    ax.tick_params('y', direction='out', pad=0.4)\n",
    "\n",
    "    # want a more natural, table-like display\n",
    "    ax.invert_yaxis()\n",
    "    ax.xaxis.tick_top()\n",
    "\n",
    "    # remove ticks\n",
    "    ax.xaxis.set_ticks_position('none')\n",
    "    ax.yaxis.set_ticks_position('none')\n",
    "\n",
    "    # so labels of x and y do not show up and my labels show up instead\n",
    "    ax.set_xticklabels(list(dataset.columns), fontsize=6.5,\n",
    "                       fontname=\"Arial\", color='k', minor=False)\n",
    "    ax.set_yticklabels(temp_kwargs['neworder_aminoacids'],\n",
    "                       fontsize=6.5, fontname=\"Arial\", color='k', minor=False)\n",
    "\n",
    "    # align the labels of the y axis\n",
    "    for ylabel in ax.get_yticklabels():\n",
    "        ylabel.set_horizontalalignment('center')\n",
    "\n",
    "    # _____________________________________________________________________________\n",
    "\n",
    "    # for color bar format\n",
    "    cb = plt.colorbar(heatmap, fraction=0.025, pad=0.05, aspect=5, ticks=[temp_kwargs['colorbar_scale'][0], temp_kwargs['colorbar_scale'][1]],\n",
    "                      orientation='vertical')\n",
    "    cb.ax.set_yticklabels(cb.ax.get_yticklabels(),\n",
    "                          fontsize=7, fontname=\"Arial\", color='k')\n",
    "    cb.update_ticks()\n",
    "    plt.text(len(dataset.columns)+1.2*coeff, len(dataset.columns)/2.5, 'R',\n",
    "             horizontalalignment='center', fontsize=7, fontname=\"Arial\", color='k')\n",
    "\n",
    "    # for putting title on graph\n",
    "    plt.title(temp_kwargs['title'], horizontalalignment='center',\n",
    "              fontname=\"Arial\", fontsize=10, pad=10)\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Individual correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.340371Z",
     "start_time": "2020-10-06T23:17:05.324718Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.565004Z",
     "iopub.status.busy": "2020-10-11T02:23:58.564771Z",
     "iopub.status.idle": "2020-10-11T02:23:58.575759Z",
     "shell.execute_reply": "2020-10-11T02:23:58.574655Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.564986Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_individual_correlation(self, output_file: Union[None, str, Path] = None,\n",
    "                                **kwargs):\n",
    "    '''\n",
    "    Generates a bar plot of the correlation of each amino acid mutational \n",
    "    profile (row of the heatmap) with the rest of amino acids (rows)\n",
    "    \n",
    "    Parameters\n",
    "    -----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                    \n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "\n",
    "    '''\n",
    "    # Load parameters\n",
    "    code_kwargs._parameters()\n",
    "\n",
    "    # Update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (3.5, 2))\n",
    "    temp_kwargs['yscale'] = kwargs.get('yscale', (0, 1))\n",
    "\n",
    "    # Get data\n",
    "    if '*' in temp_kwargs['neworder_aminoacids']:\n",
    "        temp_kwargs['neworder_aminoacids'].remove('*')\n",
    "    df = _calculate_correlation(\n",
    "        self.dataframe, temp_kwargs['neworder_aminoacids']).mean()**2\n",
    "\n",
    "    # Make figure\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    ticks = np.arange(0, len(df))  # label locations\n",
    "    width = 0.5\n",
    "    labels = temp_kwargs['neworder_aminoacids']\n",
    "    # Plot figure\n",
    "    ax.bar(ticks, df, width, color='blue', ec='k',)\n",
    "\n",
    "    # graph parameters\n",
    "    ax.set_xticks(ticks)\n",
    "    ax.set_xticklabels(labels, fontsize=9, fontname=\"Arial\",\n",
    "                       color='k', minor=False, rotation=0)\n",
    "    ax.set_ylabel(r'$R^2$', fontsize=10, fontname=\"Arial\",\n",
    "                  color='k', labelpad=12, rotation=0)\n",
    "    ax.set_ylim(temp_kwargs['yscale'])\n",
    "    plt.title(temp_kwargs['title'], horizontalalignment='center',\n",
    "              fontname=\"Arial\", fontsize=10, pad=5)\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Group correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.361657Z",
     "start_time": "2020-10-06T23:17:05.343614Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.578840Z",
     "iopub.status.busy": "2020-10-11T02:23:58.578585Z",
     "iopub.status.idle": "2020-10-11T02:23:58.595511Z",
     "shell.execute_reply": "2020-10-11T02:23:58.594963Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.578812Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_group_correlation(self, r2, groups=['DEHKR', 'QN', 'CASTG', 'ILMV', 'WYF'],\n",
    "                        output=False, output_file: Union[None, str, Path] = None,\n",
    "                           **kwargs):\n",
    "    '''\n",
    "    Determines which amino acids better represent the heatmap. Requires logomaker package.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    r2 : float\n",
    "        cutoff of the r**2 correlation value. Only values above that will be plot at the sequence logo\n",
    "    \n",
    "    groups : list, default ['DEHKR','QN','CASTG','ILMV','WYF']\n",
    "        groups of aa to combine together\n",
    "    \n",
    "    output : boolean, default False\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                    \n",
    "    **kwargs : other keyword arguments\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    Use logomaker to plot the most frequent residues. \n",
    "    Optional gives back the different combinations of groups and the R**2 values\n",
    "    '''\n",
    "\n",
    "    # Update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "\n",
    "    # Apply parameters\n",
    "    code_kwargs._parameters()\n",
    "\n",
    "    # If there is a stop codon, delete it\n",
    "    if '*' in temp_kwargs['neworder_aminoacids']:\n",
    "        temp_kwargs['neworder_aminoacids'].remove('*')\n",
    "        \n",
    "    # Get R2 of each combination of amino acid substitutions\n",
    "    df = _calculate_substitution_correlations(self, temp_kwargs['neworder_aminoacids'], groups)\n",
    "\n",
    "    # Filter according the the R2 correlation value\n",
    "    filtered = df.loc[df['R2'] > r2]\n",
    "    logoplot = logomaker.alignment_to_matrix(list(filtered['Combinations']))\n",
    "\n",
    "    # create Logo object\n",
    "    fig = logomaker.Logo(logoplot, font_name='Arial', color_scheme='chemistry', vpad=.1,\n",
    "                         width=.8, figsize=((len(logoplot)+1)/2.5, 1))\n",
    "\n",
    "    # style using Logo methods\n",
    "    fig.style_xticks(anchor=0, spacing=1, rotation=0)\n",
    "\n",
    "    # No yticks and no xticks (but keep labels)\n",
    "    plt.yticks([], [])\n",
    "    fig.ax.tick_params(axis='both', which='both', length=0)\n",
    "\n",
    "    # style using Axes methods\n",
    "    fig.ax.set_ylabel('Bits')\n",
    "    fig.ax.set_xlim([-0.5, len(logoplot)-0.5])\n",
    "\n",
    "    # for putting title on graph\n",
    "    plt.title(temp_kwargs['title'], horizontalalignment='center',\n",
    "              fontname=\"Arial\", fontsize=10, pad=10)\n",
    "\n",
    "    # save file, cannot save logo file for now\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "        \n",
    "    if output:\n",
    "        return df\n",
    "\n",
    "\n",
    "def _calculate_substitution_correlations(self, aminoacids, groups):\n",
    "    '''if a set of residues was chosen, how well would they represent the entire population'''\n",
    "    \n",
    "    # Get correlation values\n",
    "    corr_values = _calculate_correlation(self.dataframe, aminoacids)**2\n",
    "    corr_values.reset_index(inplace=True)\n",
    "\n",
    "    # Get combinations\n",
    "    replacement_combinations = list(itertools.product(*groups))\n",
    "\n",
    "    # Retrieve Correlation values\n",
    "    df = pd.DataFrame()\n",
    "    df['Aminoacids'] = list(itertools.chain.from_iterable(groups))\n",
    "    for combination in replacement_combinations:  # Iterate over a combination\n",
    "        temp_list = []\n",
    "        \n",
    "        # Iterate over a group of the combination\n",
    "        for group, aa_selected in zip(groups, combination):\n",
    "            for aa_nonselected in group:  # Find correlation values from correlation plot\n",
    "                if aa_nonselected == aa_selected:\n",
    "                    temp_list.append(1)\n",
    "                else:\n",
    "                    temp_list.append(_find_correlation(\n",
    "                        aa_selected, aa_nonselected, corr_values))\n",
    "        df[combination] = temp_list  # Store in df\n",
    "    return _polishdf(df)\n",
    "\n",
    "\n",
    "def _polishdf(df):\n",
    "    df_mean = df.copy()\n",
    "    df_mean = df.mean().to_frame()\n",
    "    df_mean.reset_index(drop=False, inplace=True)\n",
    "    df_mean.rename(columns={0: 'R2'}, inplace=True)\n",
    "    df_mean['Combinations'] = list(df_mean['index'].apply(lambda x: ''.join(x)))\n",
    "    df_mean.drop(columns=['index'], inplace=True)\n",
    "    return df_mean\n",
    "\n",
    "\n",
    "def _find_correlation(aa1, aa2, corr_values):\n",
    "    return float(corr_values[aa1].loc[corr_values['Aminoacid'] == aa2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.397923Z",
     "start_time": "2020-10-06T23:17:05.364410Z"
    },
    "execution": {
     "iopub.execute_input": "2020-10-11T02:23:58.596889Z",
     "iopub.status.busy": "2020-10-11T02:23:58.596525Z",
     "iopub.status.idle": "2020-10-11T02:23:58.615957Z",
     "shell.execute_reply": "2020-10-11T02:23:58.614704Z",
     "shell.execute_reply.started": "2020-10-11T02:23:58.596862Z"
    }
   },
   "outputs": [],
   "source": [
    "def plot_pca(self, mode='aminoacid', dimensions=[0, 1], adjustlabels = False,\n",
    "             output_file: Union[None, str, Path] = None,**kwargs):\n",
    "    '''\n",
    "    Generates a plot of two PCA dimensions\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    self : object from class *Screen*\n",
    "    \n",
    "    mode : list, default 'aminoacid'\n",
    "        Can also do PCA by secondary structure element if set to \"secondary\" or \n",
    "        by individual residue if set to \"individual\".\n",
    "    \n",
    "    dimensions : list, default [0,1]\n",
    "        Specify which two PCA dimensions to plot. By default PCA1 vs PCA2.\n",
    "        Max dimension is 5.\n",
    "    \n",
    "    adjustlabels : boolean, default False\n",
    "        If set to true, it will adjust the text labels so there is no overlap. It is convenient to increase\n",
    "        the size of the figure, otherwise the algorithm will not find a solution. Requires to install adjustText package.\n",
    "    \n",
    "    output_file : str, default None\n",
    "        If you want to export the generated graph, add the path and name of the file.\n",
    "        Example: 'path/filename.png' or 'path/filename.svg'. \n",
    "                    \n",
    "    **kwargs : other keyword arguments\n",
    "        random_state : int, default 554\n",
    "    Returns\n",
    "    ----------\n",
    "    None.\n",
    "\n",
    "    '''\n",
    "\n",
    "    # load parameters\n",
    "    code_kwargs._parameters()\n",
    "\n",
    "    # update kwargs\n",
    "    temp_kwargs = copy.deepcopy(code_kwargs.kwargs())\n",
    "    temp_kwargs.update(kwargs)\n",
    "    temp_kwargs['figsize'] = kwargs.get('figsize', (2, 2))\n",
    "\n",
    "    # calculate correlation heatmap. Choose mode\n",
    "    dataset = self.dataframe.copy()\n",
    "    if mode == 'aminoacid':\n",
    "        if '*' in temp_kwargs['neworder_aminoacids']:\n",
    "            temp_kwargs['neworder_aminoacids'].remove('*')\n",
    "        dataset = _calculate_correlation(\n",
    "            dataset, temp_kwargs['neworder_aminoacids'])\n",
    "        textlabels = temp_kwargs['neworder_aminoacids']\n",
    "    elif mode == 'secondary':\n",
    "        dataset = _calculate_correlation_bysecondary(\n",
    "            dataset, self.secondary_dup)\n",
    "        textlabels = list(dataset.columns)\n",
    "    elif mode == 'individual':\n",
    "        dataset = _calculate_correlation_byresidue(dataset)\n",
    "        textlabels = list(dataset.columns)\n",
    "    \n",
    "    # plot using plot_clusters\n",
    "    dimensionstoplot, variance = _calculate_clusters(dataset, dimensions, temp_kwargs['random_state'])\n",
    "\n",
    "    # x and y\n",
    "    x = dimensionstoplot.iloc[:, 0]\n",
    "    y = dimensionstoplot.iloc[:, 1]\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=temp_kwargs['figsize'])\n",
    "    ax.scatter(x, y, s=4, c='k')\n",
    "\n",
    "    # labels\n",
    "    plt.xlabel('PCA ' + str(dimensions[0]+1) + ': ' + str(int(variance[dimensions[0]]*100))+'%',\n",
    "               fontsize=10, labelpad=5, fontweight='normal')\n",
    "    plt.ylabel('PCA ' + str(dimensions[1]+1) + ': ' + str(int(variance[dimensions[1]]*100))+'%',\n",
    "               fontsize=10, labelpad=-2, fontweight='normal')\n",
    "\n",
    "    # label of data points\n",
    "    texts = _auto_text(x, y, textlabels)\n",
    "    if adjustlabels is True:\n",
    "        adjust_text(texts, autoalign='xy')\n",
    "    \n",
    "    # set title\n",
    "    plt.title(temp_kwargs['title'], horizontalalignment='center',\n",
    "              fontname=\"Arial\", fontsize=10, pad=5)\n",
    "\n",
    "    # save file\n",
    "    code_utils._save_work(fig, output_file, temp_kwargs)\n",
    "\n",
    "    if temp_kwargs['show']:\n",
    "        plt.show()\n",
    "    return\n",
    "\n",
    "\n",
    "def _auto_text(x, y, textlabels):\n",
    "    '''auto anotates text labels'''\n",
    "    texts = [plt.annotate(textlabels[i],  # this is the text\n",
    "                          (x[i], y[i]),  # this is the point to label\n",
    "                          textcoords=\"offset points\",  # how to position the text\n",
    "                          xytext=(2, 2),  # distance from text to points (x,y)\n",
    "                          fontsize=8,\n",
    "                          ha='center')  # horizontal alignment can be left, right or center\n",
    "             for i in range(len(textlabels))]\n",
    "    return texts\n",
    "\n",
    "\n",
    "def _calculate_clusters(dataset, dimensions, random_state):\n",
    "    '''input the dataframe that needs to be correlated, the dimensions, and will calculate PCA descomposition. '''\n",
    "\n",
    "    # call pca model\n",
    "    pca = PCA(n_components=6, random_state = random_state)\n",
    "\n",
    "    # fit model to df. use aux function correlation_aminoacids\n",
    "    model = pca.fit(dataset)\n",
    "\n",
    "    # create df with PCA data\n",
    "    df_aa = pd.DataFrame((model.components_).T, columns=[\n",
    "                         'PCA1', 'PCA2', 'PCA3', 'PCA4', 'PCA5', 'PCA6'])\n",
    "\n",
    "    # use kmeans to cluster the two dimensions and color\n",
    "    dimensionstoplot = df_aa.iloc[:, np.r_[dimensions[0], dimensions[1]]]\n",
    "\n",
    "    return dimensionstoplot, pca.explained_variance_ratio_\n",
    "\n",
    "\n",
    "def _grouby_secondary(df, secondary):\n",
    "    '''\n",
    "    Groups each secondary motif and makes the mean.\n",
    "\n",
    "    Returns dataframe. Returns copy\n",
    "    '''\n",
    "    df = df.copy()\n",
    "    df.insert(4, 'Secondary', secondary)\n",
    "    df = df.groupby(['Secondary', 'Aminoacid'], as_index=False).mean()\n",
    "    df = df.loc[df['Secondary'].str.startswith(('β', 'α'))]\n",
    "    return df\n",
    "\n",
    "\n",
    "def _calculate_correlation_bysecondary(df, secondary):\n",
    "    dataset = _grouby_secondary(df, secondary)\n",
    "    dataset = dataset.pivot_table(\n",
    "        values='Score', index='Secondary',  columns='Aminoacid')\n",
    "    dataset = dataset.T.corr()\n",
    "\n",
    "    return dataset\n",
    "\n",
    "def _calculate_correlation(df, order_aminoacids):\n",
    "\n",
    "    dataset = df.copy()\n",
    "    dataset = dataset.pivot_table(\n",
    "        values='Score', index='Position',  columns='Aminoacid')\n",
    "    dataset = dataset.corr()\n",
    "    dataset = dataset.reindex(index=order_aminoacids)[order_aminoacids]\n",
    "\n",
    "    return dataset\n",
    "\n",
    "\n",
    "def _calculate_correlation_byresidue(df):\n",
    "\n",
    "    dataset = df.copy()\n",
    "    dataset = dataset.pivot_table(\n",
    "        values='Score', index='Position',  columns='Aminoacid')\n",
    "    dataset = dataset.T.corr()\n",
    "\n",
    "    return dataset"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "100px",
    "left": "21px",
    "top": "69.2px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.4,
   "position": {
    "height": "423.4px",
    "left": "944.6px",
    "right": "20px",
    "top": "119px",
    "width": "315.4px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "block",
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
