{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-12T22:36:10.914292Z",
     "start_time": "2020-10-12T22:36:06.031777Z"
    }
   },
   "outputs": [],
   "source": [
    "# Regular libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools\n",
    "from os import path\n",
    "import os\n",
    "\n",
    "# Local imports\n",
    "try:\n",
    "    from mutagenesis_visualization.main.scripts.code_class import Screen\n",
    "    import mutagenesis_visualization.main.scripts.code_utils as code_utils\n",
    "except ModuleNotFoundError:\n",
    "    import import_notebook\n",
    "    from code_class import Screen\n",
    "    import code_utils\n",
    "    __file__ = '__main__'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Demo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-06T23:17:05.682224Z",
     "start_time": "2020-10-06T23:17:05.649184Z"
    }
   },
   "outputs": [],
   "source": [
    "def demo(figure='heatmap', show=True):\n",
    "    \"\"\"\n",
    "    Performs a demonstration of the mutagenesis_visualization software.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    figure : str, default 'heatmap'\n",
    "        There are a few example plots that can be displayed to test the package is working on your station.\n",
    "        The options are 'heatmap', 'miniheatmap', 'mean', 'kernel', 'pca'\n",
    "        'position', 'secondary_mean', 'correlation', 'individual_correlation'. \n",
    "        Check the documentation for more information.\n",
    "    \n",
    "    show : boolean, default True\n",
    "        If True, will do plt.show() for each figure.\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    None.\n",
    "    \"\"\"\n",
    "    # Use relative file import to access the data folder\n",
    "    try:\n",
    "        location = os.path.dirname(os.path.realpath(__file__))\n",
    "        my_file = os.path.join(location, '../../data', 'HRas166_RBD.csv')\n",
    "    except NameError:\n",
    "        my_file = os.path.join('../../data', 'HRas166_RBD.csv')\n",
    "\n",
    "    # Load enrichment scores\n",
    "    hras_enrichment_RBD = np.genfromtxt(my_file, delimiter=',')\n",
    "\n",
    "    # Define protein sequence\n",
    "    hras_sequence = 'MTEYKLVVVGAGGVGKSALTIQLIQNHFVDEYDPTIEDSYRKQVVIDGETCLLDILDTAGQEEYSAMRDQYMRTGEGFLCVFAINNTKSFEDIHQYREQIKRVKDSDDVPMVLVGNKCDLAARTVESRQAQDLARSYGIPYIETSAKTRQGVEDAFYTLVREIRQHKLRKLNPPDESGPG'\n",
    "\n",
    "    # Define secondary structure\n",
    "    secondary = [['L0'], ['β1'] * (9 - 1), ['L1'] * (15 - 9),\n",
    "                 ['α1'] * (25 - 15), ['L2'] * (36 - 25), ['β2'] * (46 - 36),\n",
    "                 ['L3'] * (48 - 46), ['β3'] * (58 - 48), ['L4'] * (64 - 58),\n",
    "                 ['α2'] * (74 - 64), ['L5'] * (76 - 74), ['β4'] * (83 - 76),\n",
    "                 ['L6'] * (86 - 83), ['α3'] * (103 - 86), ['L7'] * (110 - 103),\n",
    "                 ['β5'] * (116 - 110), ['L8'] * (126 - 116),\n",
    "                 ['α4'] * (137 - 126), ['L9'] * (140 - 137),\n",
    "                 ['β6'] * (143 - 140), ['L10'] * (151 - 143),\n",
    "                 ['α5'] * (172 - 151), ['L11'] * (190 - 172)]\n",
    "\n",
    "    # Create object\n",
    "    hras_RBD = Screen(\n",
    "        dataset=hras_enrichment_RBD,\n",
    "        sequence=hras_sequence,\n",
    "        secondary=secondary\n",
    "    )\n",
    "\n",
    "    if figure == 'heatmap':\n",
    "        # Create heatmap plot\n",
    "        hras_RBD.heatmap(title='H-Ras 2-166', show_cartoon=True, show=show)\n",
    "    elif figure == 'miniheatmap':\n",
    "        # Condensed heatmap\n",
    "        hras_RBD.miniheatmap(title='Wt residue H-Ras', show=show)\n",
    "    elif figure == 'mean':\n",
    "        # Mean enrichment by position\n",
    "        hras_RBD.mean(\n",
    "            figsize=[6, 2.5],\n",
    "            mode='mean',\n",
    "            show_cartoon=True,\n",
    "            yscale=[-2, 0.5],\n",
    "            title='',\n",
    "            show=show\n",
    "        )\n",
    "    elif figure == 'kernel':\n",
    "        # Plot kernel dist using sns.distplot.\n",
    "        hras_RBD.kernel(\n",
    "            histogram=True, title='H-Ras 2-166', xscale=[-2, 1], show=show\n",
    "        )\n",
    "    elif figure == 'pca':\n",
    "        # PCA by amino acid substitution\n",
    "        hras_RBD.pca(\n",
    "            dimensions=[0, 1],\n",
    "            figsize=(2, 2),\n",
    "            adjustlabels=True,\n",
    "            title='',\n",
    "            show=show\n",
    "        )\n",
    "    elif figure == 'position':\n",
    "        # Create plot for position 117\n",
    "        hras_RBD.position(\n",
    "            position=117,\n",
    "            yscale=(-1.5, 0.8),\n",
    "            figsize=(3.5, 2),\n",
    "            title='Position 117',\n",
    "            output_file=None,\n",
    "            show=show\n",
    "        )\n",
    "    elif figure == 'secondary_mean':\n",
    "        hras_RBD.secondary_mean(\n",
    "            yscale=[-1, 0],\n",
    "            figsize=[3, 2],\n",
    "            title='Mean of secondary motifs',\n",
    "            show=show\n",
    "        )\n",
    "    elif figure == 'correlation':\n",
    "        # Correlation between amino acids\n",
    "        hras_RBD.correlation(\n",
    "            colorbar_scale=[0.5, 1], title='Correlation', show=show\n",
    "        )\n",
    "    elif figure == 'individual_correlation':\n",
    "        # Explained variability by amino acid\n",
    "        hras_RBD.individual_correlation(\n",
    "            yscale=[0, 0.6],\n",
    "            title='Explained variability by amino acid',\n",
    "            output_file=None,\n",
    "            show=show\n",
    "        )\n",
    "    else:\n",
    "        raise NameError('Select a valid name for a demo figure')\n",
    "    return\n",
    "\n",
    "\n",
    "def demo_datasets():\n",
    "    \"\"\"\n",
    "    Loads example datasets so the user can play with it.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    None\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    data_dict : dictionary\n",
    "        Dictionary that contains the datasets used to create the plots on the documentation.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Use relative file import to access the data folder\n",
    "    location = os.path.dirname(os.path.realpath(__file__))\n",
    "\n",
    "    # Create dictionary where to store data\n",
    "    data_dict = {}\n",
    "\n",
    "    # Retrieve H-Ras datasets and store in dict\n",
    "    my_file = os.path.join(location, '../../data', 'HRas166_RBD.csv')\n",
    "    hras_enrichment_RBD = np.genfromtxt(my_file, delimiter=',')\n",
    "    data_dict['array_hras_RBD'] = hras_enrichment_RBD\n",
    "\n",
    "    my_file = os.path.join(location, '../../data', 'HRas166_GAPGEF.csv')\n",
    "    hras_enrichment_GAPGEF = np.genfromtxt(my_file, delimiter=',')\n",
    "    data_dict['array_hras_GAPGEF'] = hras_enrichment_GAPGEF\n",
    "\n",
    "    # Beta lactamase data\n",
    "    my_file = os.path.join(location, '../../data', 'df_bla_raw.pkl')\n",
    "    df_bla_raw = pd.read_pickle(my_file)\n",
    "    data_dict['df_bla'], sequence_bla = code_utils.parse_pivot(\n",
    "        df_bla_raw, col_data='DMS_amp_625_(b)'\n",
    "    )\n",
    "\n",
    "    # Sumo\n",
    "    my_file = os.path.join(location, '../../data', 'df_sumo1_raw.pkl')\n",
    "    df_sumo1_raw = pd.read_pickle(my_file)\n",
    "    data_dict['df_sumo1'], sequence_sumo1 = code_utils.parse_pivot(\n",
    "        df_sumo1_raw, col_data='DMS'\n",
    "    )\n",
    "\n",
    "    # MAPK1\n",
    "    my_file = os.path.join(location, '../../data', 'df_mapk1_raw.pkl')\n",
    "    df_mapk1_raw = pd.read_pickle(my_file)\n",
    "    data_dict['df_mapk1'], sequence_mapk1 = code_utils.parse_pivot(\n",
    "        df_mapk1_raw, col_data='DMS_DOX'\n",
    "    )\n",
    "\n",
    "    # UBE2I\n",
    "    my_file = os.path.join(location, '../../data', 'df_ube2i_raw.pkl')\n",
    "    df_ube2i_raw = pd.read_pickle(my_file)\n",
    "    data_dict['df_ube2i'], sequence_ube2i = code_utils.parse_pivot(\n",
    "        df_ube2i_raw, col_data='DMS'\n",
    "    )\n",
    "\n",
    "    # TAT\n",
    "    my_file = os.path.join(location, '../../data', 'df_tat.pkl')\n",
    "    data_dict['df_tat'] = pd.read_pickle(my_file)\n",
    "\n",
    "    # REV\n",
    "    my_file = os.path.join(location, '../../data', 'df_rev.pkl')\n",
    "    data_dict['df_rev'] = pd.read_pickle(my_file)\n",
    "\n",
    "    # asynuclein\n",
    "    my_file = os.path.join(location, '../../data', 'df_asynuclein.pkl')\n",
    "    data_dict['df_asynuclein'] = pd.read_pickle(my_file)\n",
    "\n",
    "    # APH\n",
    "    my_file = os.path.join(location, '../../data', 'df_aph.pkl')\n",
    "    data_dict['df_aph'] = pd.read_pickle(my_file)\n",
    "\n",
    "    # b11L5\n",
    "    my_file = os.path.join(location, '../../data', 'df_b11L5F_raw.pkl')\n",
    "    df_b11L5F_raw = pd.read_pickle(my_file)\n",
    "    data_dict['df_b11L5F'], sequence_b11L5F = code_utils.parse_pivot(\n",
    "        df_b11L5F_raw, col_data='relative_tryp_stability_score'\n",
    "    )\n",
    "\n",
    "    return data_dict\n",
    "\n",
    "\n",
    "def demo_pdbs():\n",
    "    \"\"\"\n",
    "    Loads example pdbs so the user can play with it (5p21, 1erm, 1a5r, 1nd4).\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    None\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    data_dict : dictionary\n",
    "        Dictionary that contains the pdbs used to create the plots on the documentation.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Use relative file import to access the data folder\n",
    "    location = os.path.dirname(os.path.realpath(__file__))\n",
    "\n",
    "    # Create dictionary where to store data\n",
    "    pdb_dict = {}\n",
    "\n",
    "    # H-Ras\n",
    "    pdb_dict['5p21'] = os.path.join(location, '../../data', '5p21.pdb')\n",
    "\n",
    "    # Beta lactamase data\n",
    "    pdb_dict['1erm'] = os.path.join(location, '../../data', '1erm.pdb')\n",
    "\n",
    "    # Sumo\n",
    "    pdb_dict['1a5r'] = os.path.join(location, '../../data', '1a5r.pdb')\n",
    "\n",
    "    # APH\n",
    "    pdb_dict['1nd4'] = os.path.join(location, '../../data', '1nd4.pdb')\n",
    "\n",
    "    return pdb_dict\n",
    "\n",
    "\n",
    "def demo_fasta():\n",
    "    \"\"\"\n",
    "    Loads example fasta so the user can play with it.\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    None\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    data_dict : dictionary\n",
    "        Dictionary that contains the fasta used to extract the sequence conservation.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Use relative file import to access the data folder\n",
    "    location = os.path.dirname(os.path.realpath(__file__))\n",
    "    print(location)\n",
    "    # Create dictionary where to store data\n",
    "    fasta_dict = {}\n",
    "    fasta_dict['ras'] = os.path.join(\n",
    "        location, '../../data', 'Ras_family_trimmed.fasta'\n",
    "    )\n",
    "\n",
    "    return fasta_dict"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "100px",
    "left": "21px",
    "top": "69.2px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.4,
   "position": {
    "height": "423.4px",
    "left": "944.6px",
    "right": "20px",
    "top": "119px",
    "width": "315.4px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "block",
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
